### PR TITLE
Extended canvas checks

### PR DESCRIPTION
**Pull request by jonathanKingston**
This test verifies the few cases we care about:

- Drawing a known image to the canvas
- Rendering code to the canvas
- Checking perf of pixel checking and 4k image
- Cached values don't return blank after FP code runs due to caching

Sample test runs:

Without the addon in Chrome:
image

With the addon in Chrome beta:
image

With the addon in Firefox:
image

### CODE REVIEW

File: privacy-protections/fingerprinting/canvas.html [hunk 1]
```diff
@@ -0,0 +1,32 @@
+<!DOCTYPE html>
+<html lang="en">
+<head>
+    <meta charset="UTF-8">
+    <meta name="viewport" content="width=device-width, initial-scale=1.0">
+    <title>Canvas fingerprinting test page</title>
```
Comment (kdzwinel) [inline privacy-protections/fingerprinting/canvas.html:6]: Let's keep title in sync with what we call the test on the index page.

File: privacy-protections/fingerprinting/canvas.html [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+            background-color: red;
+        }
+
+        .status-pass td:first-child {
+            background-color: green;
+        }
+    </style>
+</head>
+<body>
+    <p><a href="../../">[Home]</a> ↣ <a href="../">[Privacy Protections Tests]</a> ↣ <strong>[Fingerprinting Test Page]</strong></p>
```
Comment (kdzwinel) [inline privacy-protections/fingerprinting/canvas.html:25]: Let's add a paragraph of a general explanation what's being tested.

File: privacy-protections/fingerprinting/canvas.js [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+    // Render a random 4k canvas and then get the performance of getting the pixels
+    async () => {
+        const limit = 200;
+        const canvasElement = createCanvas(3840, 2160);
+        const canvasContext = canvasElement.getContext('2d');
+        const before = performance.now();
+        const canvasData = canvasContext.getImageData(0, 0, canvasElement.width, canvasElement.height);
+        const after = performance.now();
+        const time = after - before;
+        ok(time < limit, `Getting image data must be under ${limit}ms (${time})`);
```
Comment (kdzwinel) [inline privacy-protections/fingerprinting/canvas.js:37]: Let's add a bit more context to each test
```suggestion
        ok(time < limit, `Canvas performance: getting empty canvas image data must be under ${limit}ms (${time})`);
```

File: privacy-protections/fingerprinting/canvas.js [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+        for (let i = 0; i < canvasData.data.length; i++) {
+            canvasData.data[i] = Math.floor(Math.random() * 256);
+        }
+        canvasContext.putImageData(canvasData, 0, 0);
+
+        const before2 = performance.now();
+        const canvasData2 = canvasContext.getImageData(0, 0, canvasElement.width, canvasElement.height);
+        const after2 = performance.now();
+        const time2 = after2 - before2;
+        ok(time2 < limit, `Getting image data must be under ${limit}ms (${time2})`);
```
Comment (kdzwinel) [inline privacy-protections/fingerprinting/canvas.js:48]:
```suggestion
        ok(time2 < limit, `Canvas performance: getting random canvas image data must be under ${limit}ms (${time2})`);
```

File: privacy-protections/fingerprinting/canvas.js [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+
+        // Validate canvas caching didn't return blank pixels here
+        // We can't check for equality as the browser may modify on put.
+        ok(canvasData2.data.find((v) => v !== 0), 'Data in image must not be blank');
```
Comment (kdzwinel) [inline privacy-protections/fingerprinting/canvas.js:52]:
```suggestion
        ok(canvasData2.data.find((v) => v !== 0), 'Canvas performance: caching doesn't cause image to be blank');
```

File: privacy-protections/fingerprinting/canvas.js [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+        });
+        image.onload = (e) => {
+            const refImage = e.target;
+            const canvasElement = createCanvas(refImage.width, refImage.height);
+            const canvasContext = canvasElement.getContext('2d');
+            canvasContext.drawImage(refImage, 0, 0);
+            const before = performance.now();
+            const canvasData = canvasContext.getImageData(0, 0, canvasElement.width, canvasElement.height);
+            const after = performance.now();
+            eq(canvasData.data[0], 18, 'Red should be 18');
```
Comment (kdzwinel) [inline privacy-protections/fingerprinting/canvas.js:72]:
```suggestion
            eq(canvasData.data[0], 18, 'Canvas correctness: Red should be 18');
```

etc. etc. to all checks

File: privacy-protections/fingerprinting/canvas.js [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+            const time = after - before;
+            ok(time < limit, `Getting image data must be under ${limit}ms (${time})`);
+            const knownOutputUrls = [
+                // Firefox
+                'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVQImWMQEvr/HwADgAIj6OBrXgAAAABJRU5ErkJggg==',
+
+                // Chrome
+                'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAAAXNSR0IArs4c6QAAAA1JREFUGFdjEBL6/x8AA4ACI+VyxooAAAAASUVORK5CYII='
+            ];
+            ok(knownOutputUrls.includes(canvasElement.toDataURL()), 'Output conversion has expected values');
```
Comment (kdzwinel) [inline privacy-protections/fingerprinting/canvas.js:85]: there is no safari value predefined here and it fails on Safari

BTW do we know if those prefefined values are the same on non-mac's?

**Review by kdzwinel — COMMENTED** [review]
Looks good, I proposed some changes to make this page more readable w/o looking into the code.

**Comment by kdzwinel** [comment]
Oh, and since we may want to automate it, please generate a global `results` object like in other tests.

**Comment by jonathanKingston** [comment]
@kdzwinel as this is a new page I'm going to merge this to unblock other work, if we need follow up changes please let me know :)

**Pull request merged**